### ISSUE TITLE
LFX Workspace: finetuning LLMs for Rust learning

### ISSUE THREAD

**Issue by codemaster1104** [WasmEdge/WasmEdge#3495]
### Summary

# Motivation
WasmEdge is a lightweight inference runtime for AI and LLM applications. Build specialized and finetuned models for WasmEdge community. The model should be supported by WasmEdge and its applications should benefit the WasmEdge community.

In this project, we will build and compare two finetuned model for Rust coding assistance.

## Model 1
**Objective**: Rust Q/A model for PR review bot
+ **Current Dataset**: [Dataset1](https://huggingface.co/datasets/Ak1104/Rust_Code_Explanation?row=0)
+ **Current Model**:[Model1](https://huggingface.co/Ak1104/Rust_Code_Explanation_Final)

## Model 2

+ **Current Dataset**: [Dataset2](https://huggingface.co/datasets/Ak1104/Rust_QA_Final)
+ **Current Model**:[Model2](https://huggingface.co/Ak1104/RustQA)

[Collab](https://colab.research.google.com/drive/1wlCOvklww1YvACuIRrhkdFFH_vU7Hgbn?usp=sharing) : Currently used for fine tuning of both models

# Milestones

 - [x] M1 [1 week]
 - Discuss This timeline with mentors and get suggestions on my tasks and my current methods of fine tuning
 - [x] M2 [2 Weeks] 
 - Get deeper understanding of gaianet 
 - Improving quality of dataset
 - Researching on prompt templates and testing their performance
 - Researching on training parameters
 - [x] M3 [1 Week]
 - Comparison of currently tuned model with Codestral model
 - Testing RAG implementation with current model and codestral using gaianet nodes
 - [x] M4 [2 Weeks]
 - Working on API inference of Model
 - Get deeper understanding of Github PR review bot
 - Create test cases for PR reviews
 - Testing various versions of base model for fine tuning of both models
 - [x] M5 [1.5 Week]
 - Integrating the model 1 with PR review bot
 - Hosting model 2
 - Get real life feedback for further improvement in their functionality 
 - [x] M6 [1.5 Weeks]
 - Working on final optimisation and integration
 - [x] M7 [2 Weeks]
 - Bug solving
 - Working on documentation

### Appendix

Reference links
https://github.com/WasmEdge/WasmEdge/issues/3371
https://docs.gaianet.ai/intro

**Comment by codemaster1104** [comment]
@juntao @hydai Can you please take a look at my datasets and suggest what improvements should be done and how many fields are required

**Comment by juntao** [comment]
Thank you @codemaster1104! These look great! Can you

1. Provide the source material for the two datasets you created. We need citations and acknowledgments.
2. Which model did you use as the base for finetuning?
3. Provide instructions on how to finetune the model on Unsloth using the datasets you provided.

Thanks!

**Comment by juntao** [comment]
Also, since one of our end goals is to use the model for PR reviews. We should also do more research in that area including:

1. Run the Codestral model. I would recommend using a GaiaNet node for this purpose. https://huggingface.co/gaianet/Codestral-22B-v0.1-GGUF
2. Use your dataset as an RAG dataset for the GaiaNet node. You will need to format the dataset into a markdown document, and then use this tool to generate a qdrant vector snapshot for the GaiaNet node: https://docs.gaianet.ai/creator-guide/knowledge/markdown
3. Create test cases for PR reviews. That is to develop a dataset with Rust project PRs' title+description+diff_file, and then develop a prompt+RAG context to generate a "review" of this PR from either our own finetuned model or the Codestral model.

**Comment by codemaster1104** [comment]
The current datasets are created from [rust by example](https://doc.rust-lang.org/rust-by-example/)
and I have created [collab notebook](https://colab.research.google.com/drive/1wlCOvklww1YvACuIRrhkdFFH_vU7Hgbn?usp=sharing) where I have used base model as [Llama-3-8b-Instruct 4 bit quantized version](https://huggingface.co/unsloth/llama-3-8b-Instruct-bnb-4bit) for fine tuning my LLM.

I have set up gaianet and ran [codestral](https://huggingface.co/gaianet/Codestral-22B-v0.1-GGUF/blob/main/Codestral-22B-v0.1-hf-Q4_0.gguf) on it, will look into RAG implementation of it

**Comment by juntao** [comment]
Hmm, you have two datasets. The chapter content in the second set appears to be from Rust by Examples. But the questions and explanations are generated by yourself? Or from an LLM? If the latter, what prompts you used to generate them? Thanks.

**Comment by juntao** [comment]
Another thing I am wondering is whether it is better to finetune the f16 version of the model, and then re-quantize to Q4 after finetuning. Something to consider later.

**Comment by codemaster1104** [comment]
I created few example questions myself and then used chat gpt using prompt as
'''
<Content of chapter>
Generate pairs of code explanation/ ques ans from the tert above in form of

<Example code exp pairs>
'''
I do feel like current dataset is a bit weak , I feel that same questin answer pairs in multiple ways/way of asking should be added in it

**Comment by codemaster1104** [comment]
I have accordingly updated my milestones , do inform me if anything else should be added or changed 
also could I get more info on model 2 and how are we planning on using it

**Comment by juntao** [comment]
I think dataset #2 could be used for Rust learning like this: https://flows.network/learn-rust

We should experiment with 

* Use dataset #2 to finetune a model for general Rust lang QA
* Use dataset #1 to finetune the model and dataset #2 to RAG.
* Use dataset #2 to RAG a codestral model

Also, are the QAs in dataset #1 also from the Rust by Example book? Did you extract the statements from the code examples by hand?

**Comment by codemaster1104** [comment]
Yes code explanation of dataset 1 are based off of same book and I have generated them in similar manner and prompt as stated above
So according to [flows learn rust](https://flows.network/learn-rust) do we need model 2 to communicate in different languages too?

**Comment by juntao** [comment]
Let's do Rust only first.

What I am uncertain (and requires research) is where to apply finetuning and where to apply RAG.

One of the original ideas is that people can send in their code and the model will explain it. But this may not be achievable through finetuning alone.

**Comment by codemaster1104** [comment]
Sure I will look into it and test both methods for all use cases
btw by language I was referring to Chinese, Japanese as stated in website
![Screenshot from 2024-06-25 15-18-51](https://github.com/WasmEdge/WasmEdge/assets/115527374/924a3af8-dd7a-49e9-a9b1-7d384368007a)

**Comment by juntao** [comment]
Oh, let's do English for now. Thanks

**Comment by codemaster1104** [comment]
Hey, is there a way I can get access to google colab pro?
While fine tuning my credits are ending really fast and then I have to wait for 24 hours also I cant experiment with large context length in free version

**Comment by juntao** [comment]
> Is there a way I can get access to google colab pro?

Yes. But since it connects to personal Google accounts, I think you can subscribe to it and we will reimburse you through a separate invoice?

**Comment by juntao** [comment]
BTW, feel free to use our Codestral-22b node to test your prompts.

https://0x57fa64fb75d1b8c778063adcd81d99e525b6197d.us.gaianet.network/

**Comment by codemaster1104** [comment]
# Update

 - Understood the basic functionality of gaianet
 - Implemented the RAG functionality of gaianet
 - Tested with different variations of snapshots on Llama-3-8b-Instruct
             1. Converted current dataset to snapshots [d1](https://huggingface.co/Ak1104/snapshot/blob/main/cde4.snapshot.tar.gz) [d2](https://huggingface.co/Ak1104/snapshot/blob/main/rust.snapshot.tar.gz)
             2. Generated a snapshot of entire [The Rust Programming Language Book](https://doc.rust-lang.org/book/#the-rust-programming-language) using larger context length embedding model [nomic-embed-text-v1.5](https://huggingface.co/gaianet/Nomic-embed-text-v1.5-Embedding-GGUF/resolve/main/nomic-embed-text-v1.5.f16.gguf)   [hugging face link](https://huggingface.co/Ak1104/snapshot/blob/main/rpl1_book.snapshot.tar.gz)
  
 - Fine tuned Llama-3-8b-Instruct with improved prompt template [collab]()
 - Created benchmark prompts to test reliability of model, these questions involve everything from general prompts to code completion to debugging [link](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit?usp=sharing)
 - Compared the performance of models using different prompt templates context length and embedding, result can be seen on below the benchmark ques in [benchmark doc](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit?usp=sharing)


## Outcomes
Currently I am getting best results with custom Llama-3-8b-Instruct fine tuned on dataset 2 using 8k context length with Chapter and using full [The Rust Programming Language Book](https://doc.rust-lang.org/book/#the-rust-programming-language) for RAG, I have attached a snapshot of performance compared to codestral , also all my tests on different variations are updated inn [benchmark doc](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit?usp=sharing) , currently I will be testing for some more variations which I will be updating on same doc.
![Screenshot from 2024-07-08 01-06-59](https://github.com/WasmEdge/WasmEdge/assets/115527374/ce479a6d-70aa-47fb-a7e8-9d0e1d2ec6cb)

**Comment by codemaster1104** [comment]
# Update
I have started to work on local LLM integration with github PR bot and was facing some issues after setting all the parameters right which included llm_model_name, llm_ctx_size, llm_api_endpoint, trigger_phrase,github_repo, github_owner when I made a PR I got the following logs for larger models:

17/07/2024, 01:27:42

{ "level": "ERROR", "message": "LLM returned an error for commit 4e29d5d141be2488a059b49d2f769675fe5f31b9: FetchError: invalid json response body at https://0x25b3d984d5c18b672383cc4f6827af11468fa58d.us.gaianet.network/v1/chat/completions reason: Unexpected token I in JSON at position 4", "target": "github_pr_summary" }

for smaller models without RAG they seem to be working just fine [example](https://github.com/codemaster1104/IISC/pull/2#issuecomment-2233114489) (this comment was generated by llama-3-8b fine tuned with dataset 1 and enhanced prompt template with 4k context length without RAG.

To understand this further I replicated the process of the pr_summary script in a python script to get what it was outputting and got following issue when working with larger models

    akshat@akshat:~/LFX$ /bin/python /home/akshat/LFX/github_pr.py
	Failed with status code: 504
	Error JSON:
	Error response not in JSON format:
	<html>
	<head><title>504 Gateway Time-out</title></head>
	<body>
	<center><h1>504 Gateway Time-out</h1></center>
	</body>
	</html>

It seems like for larger models I am constantly running into server timed out issue not only in pr_bot script but also when I run these model in using llamaedge and gaianet nodes.
I have been doing some debugging for this but cant find something concrete as for normal API request using curl in ubuntu terminal it seems to be working.

**Comment by codemaster1104** [comment]
and it seems to be only working on terminal as in gaianet UI the answers doesnt show up, if I use request package in code editors and send curl request and try to read it I get server timed out

**Comment by juntao** [comment]
Thanks @codemaster1104 

Can you create GaiaNet node config files for your model + knowledge base? You can use these config files as examples:

https://github.com/GaiaNet-AI/node-configs

You will need to upload the model files and knowledge snapshot on Hugging face. Thanks. We can then run these nodes for you.

For the GitHub bot, we have a bot (Rust based) that automatically comment on PR requests. You can check it out here:

https://github.com/flows-network/github-pr-summary

**Comment by codemaster1104** [comment]
Sure I was just testing somethings with codellama model, will send you config files shortly

**Comment by juntao** [comment]
CodeLlama does not support chat conversations, right? 

So, you need to use the regular /completion API instead of /chat/completion?

**Comment by codemaster1104** [comment]
In mean time work on my Question Answer model is done and have got best results with [QA_8k_withchapter_PT model](https://huggingface.co/Ak1104/QA_8k_withChapter_PT),
What should I do next for that objective

**Comment by juntao** [comment]
> In mean time work on my Question Answer model is done and have got best results with [QA_8k_withchapter_PT model](https://huggingface.co/Ak1104/QA_8k_withChapter_PT),
> 
> What should I do next for that objective

That's excellent. Please write down a step-by-step guide on how you prepared the data and how you finetuned and quantized the model.  Thanks. 

Can I run this model without a knowledge base?

**Comment by codemaster1104** [comment]
sure you can run it without knowledge base and the results for with knowledge base and without knowledge base are present in [this doc](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit#heading=h.yy0o48bg8c74)

**Comment by codemaster1104** [comment]
> CodeLlama does not support chat conversations, right? 
> 
> So, you need to use the regular /completion API instead of /chat/completion?

Is that so?
I was still working on fine-tuning it.

If the model is quantized to gguf does it make a difference?

**Comment by juntao** [comment]
> sure you can run it without knowledge base and the results for with knowledge base and without knowledge base are present in [this doc](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit#heading=h.yy0o48bg8c74)

Please document its context size, prompt template, and default system prompt in your hugging face model card page. Thank you.

**Comment by juntao** [comment]
> > CodeLlama does not support chat conversations, right? 
> > 
> > So, you need to use the regular /completion API instead of /chat/completion?
> 
> Is that so?
> I was still working on fine-tuning it.
> 
> If the model is quantized to gguf does it make a difference?


When you finetune it, you need to use its prompt template, right? I believe it's prompt template is designed to do code completion instead of chatting. 

No, quantization would not change the model behavior in that way.

**Comment by codemaster1104** [comment]
yes I figured that during fine tuning, just wanted to check if some custom prompt template may work

**Comment by codemaster1104** [comment]
Should I create a pr for my model in https://github.com/GaiaNet-AI/node-configs ?

**Comment by codemaster1104** [comment]
this is the [config](https://github.com/codemaster1104/node-configs/tree/main/Llama-3-8b_Rust_8k_PT_RPL) of model I want to run.
If this give good result in PR_review bot I will make a pr on main repo to add that model

**Comment by juntao** [comment]
Cool! I will test this soon.

Please create a README file like other configs in this repo. Please add a note in the README to specify your 

* base model
* training dataset for finetuning
* dataset for vector snapshot 

Then, create a PR to merge upstream. Thank you!

**Comment by codemaster1104** [comment]
I have created this custom fine tuning [colab](https://colab.research.google.com/drive/1U_axgYxtxYeeZ54ifkLTEqiguv57ix-p?usp=sharing) for codestral 22b but am running out of CUDA memory due to which i am not able to test weather its working or not

**Comment by juntao** [comment]
Cool. Can you send me a clean version of the collab script so that I can simply import into my account and run it on an H-100. Thanks.

**Comment by codemaster1104** [comment]
sure, will also add gguf conversion part to this collab and will inform you then

**Comment by codemaster1104** [comment]
# Update

+ I have fine tuned llama 3.1 on my [combined dataset](https://huggingface.co/datasets/Ak1104/combined_dataset_rust): [finetuned model](https://huggingface.co/Ak1104/combined_3.1/blob/main/unsloth.Q4_K_M.gguf)
+ Fine tune llama-3-70B on same dataset : [colab](https://colab.research.google.com/drive/1A-L8JWlAd5cn7VYRJxc01fd9KokBPKj3?usp=sharing) this is the complete clean version for 70B model and also have different quantization options , just add your hugging face token and model repo in this [cell](https://colab.research.google.com/drive/1A-L8JWlAd5cn7VYRJxc01fd9KokBPKj3#scrollTo=FqfebeAdT073&line=8&uniqifier=1)
+ Improved the snapshot for RAG :[snapshot](https://huggingface.co/Ak1104/snapshot/blob/main/rbe.snapshot.tar.gz) 

## Issues
1. Currently llamaedge doesn't support llama 3.1 , but this [PR](https://github.com/LlamaEdge/LlamaEdge/pull/201) may solve it.
2. there has been some recent changes in llama.cpp and gguf quantization function of it due to which recently fine tuned models are returning gibberish when running on llamaedge whereas when same model is being inferenced on collab or using any other service like GPT4All the inference is working fine, I have been trying to get to the roots of this issue but currently I am not entirely sure of its exact cause.

**Comment by juntao** [comment]
Thanks. 

LlamaEdge does support llama 3.1. The PR you referred to was about the tool use prompt template of llama 3.1.

**Comment by juntao** [comment]
While trying to figure the finetuned model issues, can you try the RAG dataset on llama 3.1 and Codestral? Would like to see how they perform.

**Comment by codemaster1104** [comment]
[colab](https://colab.research.google.com/drive/1cUu7M8gFuG077Roma2WjoRiOcb0IaG6L?usp=sharing) : This is the complete clean version of [codestral-22B](https://huggingface.co/mistralai/Codestral-22B-v0.1) fine tuning with q4 gguf quantization.
I recommend running it with about 100 gigs of ram and about 200 gigs of disk space.
Currently this colab has my huggingface token and saves the model in my id which can be changed from this [cell](https://colab.research.google.com/drive/1cUu7M8gFuG077Roma2WjoRiOcb0IaG6L#scrollTo=baMRQR_-iIeC&line=2&uniqifier=1)

Do let me know if there are any issues while running this.

**Comment by codemaster1104** [comment]
> While trying to figure the finetuned model issues, can you try the RAG dataset on llama 3.1 and Codestral? Would like to see how they perform.

I have added [llama 3.1 8B Q5_K_M](https://huggingface.co/gaianet/Meta-Llama-3.1-8B-Instruct-GGUF/blob/main/Meta-Llama-3.1-8B-Instruct-Q5_K_M.gguf) responses with and without RAG on [benchmark doc](https://docs.google.com/document/d/1w3nSz_V-NBkdf369a-m5UgvMKMFqiWDyh2wvfrEl7Yo/edit) starting from page 35

For codestral I was not able to generate output on my system due to hardware limitations so unfortunately you would have to test it on your system

I will now get started with documentation of fine tuning

**Comment by codemaster1104** [comment]
# Update
I have added all the instructions and tutorials in both [finetuning with unsloth](https://colab.research.google.com/drive/1wlCOvklww1YvACuIRrhkdFFH_vU7Hgbn) and [finetuning without unsloth](https://colab.research.google.com/drive/1cUu7M8gFuG077Roma2WjoRiOcb0IaG6L#scrollTo=vyBRXr8_dSmr) incase someone tries to follow this in future.
As discussed I have improved the RAG database : [Link](https://huggingface.co/Ak1104/snapshot/blob/main/final.snapshot.tar.gz)

**Comment by juntao** [comment]
Can you provide the raw data table

* text for vectorization
* text for retrieved context

for the RAG dataset?

I also want to make sure that the RAG dataset incorporates all 3 books and the finetune QAs you created (as we no longer use the finetuned models). Thanks.

**Comment by codemaster1104** [comment]
Sure, 
here is the best version of [snapshot](https://huggingface.co/Ak1104/snapshot/blob/main/final2.snapshot.tar.gz) and it was generated from [this txt](https://huggingface.co/Ak1104/snapshot/blob/main/final.txt)

I have also created a function for retrieving context and text for vectorization from any raw data automatically [here](https://github.com/codemaster1104/embedding-tools/blob/main/paragraph_embed/generate_para_vectors.py) in paragraph_embed utility.

For my current dataset it consists of formatted data from book and my data which i used to fine-tune model which was generated from book [rust by example](https://doc.rust-lang.org/rust-by-example/) in similar manner to the function i have created in paragraph_embed

**Comment by juntao** [comment]
Just to confirm that the source text for 

https://huggingface.co/Ak1104/snapshot/blob/main/final.txt

is from the "Rust by Example" book, right? How about content from the "Rust Programming Language" book? I think we need them in the knowledge base as well. Thanks.

**Comment by codemaster1104** [comment]
> Just to confirm that the source text for 
> 
> https://huggingface.co/Ak1104/snapshot/blob/main/final.txt
> 
> is from the "Rust by Example" book, right? How about content from the "Rust Programming Language" book? I think we need them in the knowledge base as well. Thanks.
> 

It has content from both books and in addition it has Q/A pairs generated from rust by example

**Comment by hydai** [comment]
Hi @juntao and @codemaster1104 
Just want to check if this issue is resolved or not.

**Comment by codemaster1104** [comment]
Hey @hydai , yes this project was completed
